### ISSUE TITLE
FFT/Wavelet look-ahead bias

### ISSUE THREAD

**Issue by windwine** [borisbanushev/stockpredictionai#19]
Hi Boris, 

Thanks for the nice walk through. One of my concerns is that when you use the full sample (1:T) to construct the "trends" you have implicitly used future info for data point at time t. DFT/wavelet is based on the whole sample and in reality at time t you should not be able to get information from t+1:T to construct your Xs. You will inventively have very high forecast precision. 

Thanks.

Regards,

Qi

**Comment by tomzo** [comment]
I had the same impression based on the short FFT snippet, it looks like the transformation was used on entire dataset.
Here is a simple illustration showing how FFT encodes information from the future in the current value.
![fft_cheat](https://user-images.githubusercontent.com/5710970/53655319-1e9a7500-3c50-11e9-8226-1f38de66a40c.png)
Notice that original process contains following values:
```
0 0 0 0 1 1 1 1
```
Yet before index 3 you can already determine that next value will be somewhere higher. Which is obviously impossible, because process was constant until then.

**Comment by dpsugasa** [comment]
Hi Tomasz, great illustration. How would you amend the code to eliminate
this problem?

**Comment by tomzo** [comment]
If the FFT was supposed to be as input, it would suffice to use it as a *rolling* statistic, very much like other MAs, Std dev, etc. 
Just like @windwine said, it is enough to use slice of current data up to current index `t` when computing fft. A new fft would have to be computed at each point in time. Either with `0:t` or `t-fft_period:t` slice used as input.

**Comment by windwine** [comment]
Yes, if you want a pure realistic setup you have to use a rolling/expanding slice to use any of these non-causal filter. One thing bothers/pisses me off is that many of the top AI/Finance journals are using the whole sample to get PCA/FFT/EMD... then use those lookahead-bias features to do the regular training/validation/testing ML tasks. They all show fantastic strategy results but actually those professors are not driving a Lamborghini. Those top journals have to get some better reviewers, otherwise it is just wasting us time to read those paper with high hope:-)

**Comment by dpsugasa** [comment]
Very good thoughts guys. If one wanted to us FFT to estimate trend (whether
you were using it as input or not) would it be appropriate to just remove
the last observation in the data set?

**Comment by IISuperluminaLII** [comment]
yes. VAE will still work fine with 111 features instead of 112

**Comment by pete21** [comment]
> would it be appropriate to just remove the last observation in the data set?

I believe no. Neither input dataset, nor output dataset. FFT will theoretically, if I remember correctly, look infinitely out into past and future. The solution is a rolling window calculation I guess.
